backend: headlamp: Add OIDC PKCE support

k-airos opened the pull request:
## Summary

Add PKCE support for OIDC authentication flow

## Related Issue

Fixes #3137

## Changes

- Added PKCE configuration: Newflag (default: false) to enable/disable PKCE
- Updated Config struct: Added OidcUsePKCE field to support PKCE configuration
- Implemented PKCE cryptographic functions
- Enhanced OauthConfig struct: Added field to store PKCE verifier

k-airos committed:
backend: Add PKCE support

Use the golang oauth2 PKCE helper functions.
Clean up OIDC auth config state when done with using it.
Use unique state variable instead of cluster name.
Use a mutex to protect the oauthRequestMap from races.

Co-authored-by: René Dudfield <renedudfield@microsoft.com>

illume commented:
Thanks @k-airos !

I asked in the issue for feedback, and ran the CI checks.

We'd need to documentation into the headlamp helm chart at some point (README.md, values.yaml and values.schema. Let me know if you'd like to add it in this PR, or we can do it after this is merged.
- https://github.com/kubernetes-sigs/headlamp/blob/main/charts/headlamp/values.yaml
- https://github.com/kubernetes-sigs/headlamp/blob/main/charts/headlamp/values.schema.json
- https://github.com/kubernetes-sigs/headlamp/blob/main/charts/headlamp/README.md

k-airos committed:
helm: Add pkce support to the helm and readme docs

k-airos commented:
Thanks! I've added the documentation to README.md, updated values.yaml and values.schema.json accordingly.
Let me know if anything else should be adjusted!

illume commented:
@k-airos much appreciated.

I see Erik from the issue gave a thumbs up. Let's wait a little bit to see if anyone else from that issue has feedback.

illume commented:
@yolossn @ashu8912 When you get a chance, can you please review?

yolossn reviewed (commented):
I tried testing this in in-cluster and the app doesn't work without the client-secret because of this client-secret check. We have to make client-secret optional in the kubconfig.OidcConfig for the user to make headlamp work without client-secret.

https://github.com/kubernetes-sigs/headlamp/blob/68e74773825f910c28ff3f4401c1dd5ee6683646/backend/pkg/kubeconfig/kubeconfig.go#L911

illume commented:
> I tried testing this in in-cluster and the app doesn't work without the client-secret because of this client-secret check. We have to make client-secret optional in the kubconfig.OidcConfig for the user to make headlamp work without client-secret.

I made the client-secret optional in the kubconfig.OidcConfig in fb511913100476751b9ef89d1e8e9fcaa628e50e

illume commented:
I added a test case for the PKCE helm config, and fixed the error copilot pointed out.

illume commented:
- Added the funlen lint ignore for the lint error in the big config function
- Did some more review, and made a couple more notes about tests requires and about a memory leak

illume commented on charts/headlamp/values.yaml:
```diff
@@ -78,6 +78,8 @@ config:
     # -- Use 'access_token' instead of 'id_token' when authenticating using OIDC
     useAccessToken: false
+    # -- Use PKCE (Proof Key for Code Exchange) for enhanced security in OIDC flow
+    usePKCE: false
```
Updated the PR description.

illume reviewed (approved):
🎉 thanks!
